### PR TITLE
Remove character limit in CEdit

### PR DESCRIPTION
**Pull request by MatiRg**
Edit Control now can use infinite character limit and of course it can still be set to use certain amount of characters.
This issue:
https://github.com/colobot/colobot/issues/820
and I think also:
https://github.com/colobot/colobot/issues/300
In future You probally will migrate to CEGUI so all UI issues will be then resolved?

### CODE REVIEW

File: src/ui/controls/edit.h [hunk 1]
```diff
@@ -232,51 +229,51 @@ class CEdit : public CControl
...
-    Math::Point     m_mouseFirstPos;
-    Math::Point     m_mouseLastPos;
-    float       m_column;
-
-    bool        m_bCapture;
-
-    bool        m_bUndoForce;
-    OperUndo    m_undoOper;
-    EditUndo    m_undo[EDITUNDOMAX];
+    std::unique_ptr<CScroll>                 m_scroll;           // vertical scrollbar on the right
```
Comment (krzys-h) [inline src/ui/controls/edit.h]: Please avoid unnecessary formatting changes like these in pull requests, they make them harder to review. Besides, aligning to long names like `std::array<HyperHistory, EDITHISTORYMAX>` decreases readability a lot.

File: src/ui/controls/edit.h [hunk 2]
```diff
@@ -132,8 +124,13 @@ class CEdit : public CControl
...
 
-    void        SetMaxChar(int max);
-    int         GetMaxChar();
+    void        SetMaxChar(int max);
+    //! Get max characters that can be hold before truncation or if is Infinite edit actual allocated characters in m_text
+    int         GetMaxChar();
+    //! Has no character limit for edit
+    bool        IsInfinite();
+    //! Set no character limit for edit
+    void        SetInfinite();
```
Comment (krzys-h) [inline src/ui/controls/edit.h]: I don't think having a separate function for this is really necessary. Infinite should be the default, and if you ever need to revert SetMaxChar setting, you can always just do `SetMaxChar(std::numeric_limits<int>::max())`

File: src/ui/controls/edit.h [hunk 3]
```diff
@@ -232,51 +229,51 @@ class CEdit : public CControl
...
-
-    bool        m_bCapture;
-
-    bool        m_bUndoForce;
-    OperUndo    m_undoOper;
-    EditUndo    m_undo[EDITUNDOMAX];
+    std::unique_ptr<CScroll>                 m_scroll;           // vertical scrollbar on the right
+
+    int                                      m_maxChar;          // max length of the buffer m_text or -1 to be infinite
+    std::vector<char>                        m_text;             // text (without zero terminator)
```
Comment (krzys-h) [inline src/ui/controls/edit.h]: Wouldn't it be way easier to just make this (and m_format) an std::string?

File: src/object/implementation/program_storage_impl.cpp [hunk 4]
```diff
@@ -131,7 +131,7 @@ Program* CProgramStorageObjectImpl::CloneProgram(Program* program)
 
     // TODO: Is there any reason CScript doesn't have a function to get the program code directly?
     auto edit = MakeUnique<Ui::CEdit>();
-    edit->SetMaxChar(Ui::EDITSTUDIOMAX);
+    edit->SetInfinite();
```
Comment (krzys-h) [inline src/object/implementation/program_storage_impl.cpp]: Not really necessary as infinite is now the default
(also applies to all other occurrences of this but I won't copy-paste this comment a billion times)

File: src/ui/controls/edit.cpp [hunk 5]
```diff
@@ -47,19 +47,20 @@
...
-const float DELAY_SCROLL    = 0.1f;
+const float DELAY_SCROLL     = 0.1f;
 //! expansion for \b;
-const float BIG_FONT        = 1.6f;
-
-
-
+const float BIG_FONT         = 1.6f;
+//! start characters
+const int DEFAULT_CHARACTERS = 100;
```
Comment (krzys-h) [inline src/ui/controls/edit.cpp]: This constant is not really needed, default is infinite anyway

File: src/ui/controls/edit.cpp [hunk 6]
```diff
@@ -1256,14 +1260,25 @@ void CEdit::DrawColor(Math::Point pos, Math::Point dim, Gfx::Color color)
 // Give the text to edit.
 
 void CEdit::SetText(const char *text, bool bNew)
-{
+{
     int     i, j, font;
     bool    bBOL;
+
+    if ( text == nullptr )
```
Comment (krzys-h) [inline src/ui/controls/edit.cpp]: For handling of wrong function arguments it's a lot better to use `assert()` than to fail silently.

File: src/ui/controls/edit.cpp [hunk 7]
```diff
@@ -1256,14 +1260,25 @@ void CEdit::DrawColor(Math::Point pos, Math::Point dim, Gfx::Color color)
...
     bool    bBOL;
+
+    if ( text == nullptr )
+    {
+        return;
+    }
 
     if ( !bNew )  UndoMemorize(OPERUNDO_SPEC);
 
     m_len = strlen(text);
```
Comment (krzys-h) [inline src/ui/controls/edit.cpp]: Try convert all C style string functions to std::string equivalents whenever possible (in this case, also change function argument to std::string)

File: src/ui/controls/edit.cpp [hunk 8]
```diff
@@ -1439,414 +1455,420 @@ void CEdit::FreeImage()
 
 bool CEdit::ReadText(std::string filename, int addSize)
 {
-    int         len, i, j, n, font, iLines, iCount;
```
Comment (krzys-h) [inline src/ui/controls/edit.cpp]: You did something weird to whitespace in this whole block (and a few other ones) but I can't quite figure out what. Indentation seems correct, line endings possibly?

File: src/ui/controls/edit.cpp [hunk 9]
```diff
@@ -1939,10 +1963,23 @@ void CEdit::SetMaxChar(int max)
 }
 
 int CEdit::GetMaxChar()
-{
+{
+    if ( IsInfinite() )
+    {
+        return m_text.size();
```
Comment (krzys-h) [inline src/ui/controls/edit.cpp]: Split this into a separate function

File: src/ui/controls/edit.cpp [hunk 10]
```diff
@@ -2120,11 +2157,22 @@ void CEdit::SetMultiFont(bool bMulti)
     m_format.clear();
 
     if (bMulti)
-    {
-        m_format.reserve(m_maxChar+1);
-        for (int i = 0; i <= m_maxChar+1; i++)
-        {
-            m_format.push_back(m_fontType);
+    {
+        if ( !IsInfinite() )
```
Comment (krzys-h) [inline src/ui/controls/edit.cpp]: It's not really good to depend on the value of IsInfinite() in code like this, using std::string like mentioned in other comments should help eliminate that.

File: src/ui/controls/edit.h [hunk 11]
```diff
@@ -132,8 +124,13 @@ class CEdit : public CControl
     bool        ReadText(std::string filename, int addSize=0);
     bool        WriteText(std::string filename);
 
-    void        SetMaxChar(int max);
-    int         GetMaxChar();
+    void        SetMaxChar(int max);
+    //! Get max characters that can be hold before truncation or if is Infinite edit actual allocated characters in m_text
+    int         GetMaxChar();
+    //! Has no character limit for edit
+    bool        IsInfinite();
```
Comment (krzys-h) [inline src/ui/controls/edit.h]: This should be removed along with SetInfinite(), external code doesn't really have any reason to care whether the the CEdit instance is infinite or not, if it does you can always use GetMaxChar() instead

File: src/ui/controls/edit.cpp [hunk 12]
```diff
@@ -47,19 +47,20 @@
...
 //! expansion for \b;
-const float BIG_FONT        = 1.6f;
-
-
-
+const float BIG_FONT         = 1.6f;
+//! start characters
+const int DEFAULT_CHARACTERS = 100;
+//! infinite edit flag
+const int INFINITE_EDIT = -1;
```
Comment (krzys-h) [inline src/ui/controls/edit.cpp]: You don't really need a separate flag, just use `std::numeric_limits<int>::max()`

**Comment by MatiRg** [comment]
I will do changes when I will have time thanks.

**Comment by MatiRg** [comment]
Nevermind  I will rewrite changes to CEdit from start

**Comment by MatiRg** [comment]
std::numeric_limits<int>::max() gives bad_alloc exception so we will need lower number.

**Comment by MatiRg** [comment]
I reworked code to use std::string as You told it was way easier it`s still not fully done.

File: src/ui/studio.cpp [hunk 13]
```diff
@@ -1619,25 +1618,25 @@ bool CStudio::ReadProgram()
...
-    if ( p == nullptr || p != filename+strlen(filename)-4 )
-    {
-        strcat(filename, ".txt");
-    }
-    strcpy(dir, SearchDirectory(true).c_str());
-    strcat(dir, filename);
+    p = filename.find(".txt");
+    if ( p == std::string::npos )
+    {
+        filename += ".txr";
```
Comment (melex750) [inline src/ui/studio.cpp]: txt*

File: src/ui/studio.cpp [hunk 14]
```diff
@@ -1657,32 +1656,32 @@ bool CStudio::WriteProgram()
...
-    }
-    strcpy(dir, SearchDirectory(true).c_str());
-    strcat(dir, filename);
+    filename = pe->GetText(100);
+    if ( filename.empty() )  return false;
+
+    p = filename.find(".txt");
+    if ( p == std::string::npos )
+    {
+        filename += ".txr";
```
Comment (melex750) [inline src/ui/studio.cpp]: txt*

**Comment by MatiRg** [comment]
ok I`am on it

**Comment by MatiRg** [comment]
I still need to deal with CRLF endings

File: src/ui/controls/edit.cpp [hunk 15]
```diff
@@ -90,11 +89,10 @@ bool IsSep(int character)
 //! Object's constructor.
 CEdit::CEdit()
     : CControl(),
+      m_text( CHARACTERS_MAX + 1, '\0' ),
```
Comment (krzys-h) [inline src/ui/controls/edit.cpp]: You are allocating 1MB of memory every time CEdit is created, this is effectively no better than an array with hardcoded size. Just leave it as an empty std::string object. Then you should be able to remove the CHARACTERS_MAX constant completly.

File: src/ui/controls/edit.cpp [hunk 16]
```diff
@@ -1940,7 +1936,7 @@ void CEdit::SetMaxChar(int max)
 
 int CEdit::GetMaxChar()
 {
-    return m_maxChar;
+    return m_text.capacity() - 1;
```
Comment (krzys-h) [inline src/ui/controls/edit.cpp]: m_text.capacity() can't really be used to determine the max amount of characters that CEdit is allowed to hold as it's value can be automatically adjusted by the STL implementation. This information should still be held as separate variable. See http://www.cplusplus.com/reference/string/string/capacity/

> Notice that this capacity does not suppose a limit on the length of the string. When this capacity is exhausted and more is needed, it is automatically expanded by the object (reallocating it storage space).
> 
> The capacity of a string can be altered any time the object is modified, even if this modification implies a reduction in size or if the capacity has not been exhausted (this is in contrast with the guarantees given to capacity in vector containers).

File: src/ui/studio.cpp [hunk 17]
```diff
@@ -512,30 +512,30 @@ void CStudio::SearchToken(CEdit* edit)
             text = const_cast<char *>(GetObjectName(type));
             if ( text[0] != 0 )
             {
-                if ( strcmp(token, text) == 0 )
+                if ( token == text ) //strcmp(token, text) == 0 )
```
Comment (krzys-h) [inline src/ui/studio.cpp]: remove the commented out part

File: src/ui/studio.cpp [hunk 18]
```diff
@@ -512,30 +512,30 @@ void CStudio::SearchToken(CEdit* edit)
...
-                    SetInfoText(std::string(token), true);
+                    SetInfoText(token, true);
                     return;
                 }
             }
             text = const_cast<char *>(GetObjectAlias(type));
             if ( text[0] != 0 )
             {
-                if ( strcmp(token, text) == 0 )
+                if ( token == text ) //strcmp(token, text) == 0 )
```
Comment (krzys-h) [inline src/ui/studio.cpp]: remove the commented out part

File: src/ui/controls/edit.cpp [hunk 19]
```diff
@@ -1457,23 +1453,25 @@ bool CEdit::ReadText(std::string filename, int addSize)
...
 
-    m_maxChar = len+addSize+100;
     m_len = len;
     m_cursor1 = 0;
     m_cursor2 = 0;
 
     FreeImage();
 
-    m_text = std::vector<char>(m_maxChar+1, '\0');
+    m_text = std::string(len2+1, '\0');
```
Comment (krzys-h) [inline src/ui/controls/edit.cpp]: You are wasting RAM again, see the comment in CEdit constructor

File: src/ui/controls/edit.cpp [hunk 20]
```diff
@@ -2121,8 +2117,8 @@ void CEdit::SetMultiFont(bool bMulti)
 
     if (bMulti)
     {
-        m_format.reserve(m_maxChar+1);
-        for (int i = 0; i <= m_maxChar+1; i++)
+        m_format.reserve(GetMaxChar());
+        for (int i = 0; i <= static_cast<int>(GetMaxChar()); i++)
```
Comment (krzys-h) [inline src/ui/controls/edit.cpp]: Since default character limit is to be assumed infinite (well, as close to infinite as possible in 32-bit numbers), you can't just start filling up infinite amount of memory. It should be dynamic like m_text, and their lengths should be kept in sync (which is probably the most complicated part of code required to fully remove the limit).

Maybe it would be a bit easier if you changed type of m_format to `std::basic_string<Gfx::FontMetaChar>`?

File: src/ui/controls/edit.cpp [hunk 21]
```diff
@@ -89,7 +89,8 @@ bool IsSep(int character)
 //! Object's constructor.
 CEdit::CEdit()
     : CControl(),
-      m_text( CHARACTERS_MAX + 1, '\0' ),
+      m_maxChar(1000000),
```
Comment (krzys-h) [inline src/ui/controls/edit.cpp]: Now that you fixed the memory allocation, you should be able to use `std::numeric_limits<int>::max()` here

**Comment by MatiRg** [comment]
Alright now it will work.

**Comment by krzys-h** [comment]
Thanks! Now I'll test it for a while to make sure there are no subtle bugs hidden somewhere as the editor is quite an important part of the game, if everything works correctly this is good to be merged.

**Comment by MatiRg** [comment]
Thanks!
